geomancer/spells/base.py

```diff
@@ -19,64 +19,123 @@
+        ------
+        NotImplementedError
+            This is an abstract class method
+        """
+        raise NotImplementedError
 
     @logger.catch
-    @abc.abstractclassmethod
-    def cast(
-        cls,
-        on,
-        df,
-        client,
-        source_table,
-        feature_name,
-        column='geometry',
-        within=10 * 1000,
-        bq_options=BQConfig,
-        **kwargs
-    ):
+    def cast(self, df, client, bq_options=BQConfig, **kwargs):
         """Apply the feature transform to an input pandas.DataFrame
 
         This is an abstract class method, and must be implemented in each subclass.
 
         Parameters
         ----------
-        on : str
-            Feature class to compare upon
         df : pandas.DataFrame
             Dataframe containing the points to compare upon. By default, we
             will look into the :code:`geometry` column. You can specify your
             own column by passing an argument to the :code:`column` parameter.
         client : google.cloud.client.Client
             Cloud Client for making requests.
-        source_table : str
-            BigQuery table to run queries against.
-        feature_name : str
-            Column name for the output feature.
-        column : str, optional
-            Column to look the geometries from. The default is :code:`geometry`
-        within : float, optional
-            Look for values within a particular range. Its value is in meters,
-            the default is :code:`10,000` meters.
         bq_options : geomancer.BQConfig
             Specify configuration for interacting with BigQuery
 
         Returns
         -------
         pandas.DataFrame
             Output dataframe with the features per given point
-
-        Raises
-        ------
-        NotImplementedError
-            This is an abstract class method
         """
-        raise NotImplementedError
+        # Load dataframe into bq with expiry
+        dataset = bq.fetch_bq_dataset(client, dataset_id=bq_options.DATASET_ID)
+        table_path = bq.upload_df_to_bq(
+            df=df,
+            client=client,
+            dataset=dataset,
+            expiry=bq_options.EXPIRY,
+            max_retries=bq_options.MAX_RETRIES,
+        )
+
+        # Retrieve table metadata
+        database_url = "bigquery://{}".format(client.project)
+        engine = create_engine(database_url)
+        metadata = MetaData(bind=engine)
+        source = Table(self.source_table, metadata, autoload=True)
+        target = Table(table_path, metadata, autoload=True)
+
+        # Build query
+        query = self.query(source, target)
+
+        # Filter
+        query = select(
+            [
+                query,
+                func.row_number()
+                .over(
+                    partition_by=query.c["__index_level_0__"],
+                    # TODO: order_by should be a parameter
+                    order_by=query.c[self.feature_name].desc(),
```

Comment:
Will this work if we want the `NumberOf` POIs `within` a certain range? The original BigQuery query looks like this:

```sql
WITH
  pairs AS (
  WITH
    pois AS (
    SELECT
      osm_id,
      pois.WKT AS WKT
    FROM
      `osm.gis_osm_pois_free_1` AS pois)
  SELECT
    ST_DISTANCE(points.WKT,
      pois.WKT) AS dist,
    points.WKT AS points_WKT,
    points.pkey AS house_pkey,
    pois.WKT AS roads_WKT,
    pois.osm_id AS osm_id
  FROM
    `houses` AS points,
    pois AS pois
  WHERE
    ST_DISTANCE(points.WKT,
      pois.WKT) < 0.5*1000)
SELECT
  house_pkey,
  ST_ASTEXT(points_WKT) AS house_WKT,
  COUNT(DISTINCT(osm_id)) AS num_pois
FROM
  pairs
GROUP BY
  house_pkey,
  house_WKT
```

Reply:
If not, I'm wondering if the whole query (including the filters) is just inside the `query()` method.

Reply:
Gets. Ok will modify